src/gpgmm/utils/ContiguousLinkNodeAllocator.h

```diff
@@ -0,0 +1,113 @@
+// Copyright 2022 The GPGMM Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GPGMM_COMMON_CONTIGUOUSLINKNODEALLOCATOR_H_
+#define GPGMM_COMMON_CONTIGUOUSLINKNODEALLOCATOR_H_
+
+#include "gpgmm/utils/Assert.h"
+#include "gpgmm/utils/LinkedList.h"
+
+#include <queue>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace gpgmm {
+
+    // ContiguousLinkNodeAllocator is an allocator designed to store objects contiguously by storing
+    // them in std::vectors of a fixed size. This class was specifically designed to be used with
+    // the LinkNode type, which represents nodes in a LinkedList. By using this allocator, LinkNodes
+    // are stored contiguously instead of in a fragmented manner, which reduces memory fragmentation
+    // and pointer chasing when iterating a LinkedList.
+    // Use placement new to allocate:
+    //     T* objectPointer = (allocator.Allocate()) T();
+    // Use the provided Deallocate to free the allocation for reuse:
+    //     allocator.Deallocate(objectPointer);
+
+    template <typename T>
+    class ContiguousLinkNodeAllocator {
+      public:
+        explicit ContiguousLinkNodeAllocator(uint64_t objectsPerBlock)
+            : mObjectsPerBlock(objectsPerBlock) {
+        }
+
+        ~ContiguousLinkNodeAllocator() {
+            for (std::vector<ContiguousLinkNodeAllocation>* blockPtr : mAllocations) {
+                delete blockPtr;
```

Comment:
nit: if you must `delete`, prefer `SafeDelete`.

Reply:
No longer needed after using a unique_ptr